06_puppy/sirigithub/mapStore.go

```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"fmt"
+)
+
+type MapStore map[int]Puppy
+
+func NewMapStore() MapStore {
+	return MapStore{}
+}
+
+/** CreatePuppy takes a Puppy with a user provided ID, and creates a new Puppy if
+the ID does not already exist in the store.
+**/
+func (m MapStore) CreatePuppy(puppy *Puppy) error {
+	if _, exists := m[puppy.ID]; exists {
+		return fmt.Errorf("puppy with ID  %d already exists", puppy.ID)
+	}
+	m[puppy.ID] = *puppy
+	return nil
+}
+
+/** UpdatePuppy overrides an existing puppy in the store,
+returns an error if id is not found or does not match the puppy ID **/
+func (m MapStore) UpdatePuppy(id int, p *Puppy) error {
+	if _, exists := m[id]; !exists {
+		return fmt.Errorf("puppy %d to update does not exist in the map", id)
+	}
+	if id != p.ID {
+		return fmt.Errorf("puppy Id %d to update does not match %d", id, p.ID)
+	}
+	m[id] = *p
+	return nil
+}
+
+/** DeletePuppy deletes an existing puppy from the store **/
+func (m MapStore) DeletePuppy(id int) (bool, error) {
```

Comment:
no need for bool and error. error alone can encode "not found" case. (successful deletion: err==nil)

Reply:
Good tip @juliaogris . Agree implemented.